vllm/v1/core/single_type_kv_cache_manager.py

```diff
@@ -684,6 +684,27 @@ def allocate_new_blocks(
             )
         return super().allocate_new_blocks(request_id, num_tokens)
 
+    def get_num_skipped_tokens(self, num_computed_tokens: int) -> int:
+        # Get the number of tokens for which the cache blocks can be freed.
+
+        # A running request needs the previous block for computing the 1st token
+        # of the subsequent block, and then the previous block can be freed.
+        # Example:
+        # block_size = 512
+        # blocks 0:[1,512] 1:[513,1024] 2:[1025, 1536] ...
+        # num_computed_tokens == 1023 -> keep from 1
+        # num_computed_tokens == 1024 -> keep from 1
+        # num_computed_tokens == 1025 -> keep from 2
+
+        # Current vLLM block freeing logic operates with num_computed_tokens:
+        # last_useful_block = self.get_num_skipped_tokens(num_computed_tokens)
+        #                     // self.block_size
+        # Solution:
```

Comment:
`num_computed_tokens - 1` is a very intuitive implementation but the above analysis is quite complex.  If you think people need help to understand this piece of code, what about adding some comments in `get_num_skipped_tokens` like
```
    def get_num_skipped_tokens(self, num_computed_tokens: int) -> int:
        """
        Get the number of tokens that will be skipped for attention computation.
        Blocks that only contain skipped tokens will be freed.

        Args:
            num_computed_tokens: The number of tokens that have been computed.

        Returns:
            The number of tokens that will be skipped for attention computation.
        """
```